src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_padding.cpp

```diff
@@ -14,6 +15,49 @@ using namespace cldnn;
 using namespace ov::intel_gpu;
 
 void prepare_padding::run(program& p) {
+    if (p.get_config().get_use_onednn()) {
+        for (const auto& node : p.get_processing_order()) {
+            if (!node->is_type<fully_connected>())
+                continue;
+
+            if (node->get_preferred_impl_type() != impl_types::onednn)
+                continue;
+
+            auto& input_node = node->get_dependency(1);
+            if (input_node.is_type<data>()
+                && input_node.is_constant()
+                && !input_node.is_dynamic()
+                && (input_node.get_output_layout(0).data_type == cldnn::data_types::u4
+                    || input_node.get_output_layout(0).data_type == cldnn::data_types::i4)) {
+                const size_t alignment = 2;
+                auto input_layout = input_node.get_output_layout(0);
+                const auto const_shape = input_layout.get_partial_shape().to_shape();
+                // Get the innermost index after trimming trailing elements in the shape.
+                size_t inner_most_idx = const_shape.size() - 1;
+                for (; inner_most_idx >= 0; --inner_most_idx) {
+                    if (const_shape[inner_most_idx] != 1)
```

Comment:
Why this check is needed? Did you face any case that innermost was not either of K or N dim? 
What if the original weight shape is K = 1 & N = 128 and K dim is innermost?

Reply:
No, the original shape is [4, 24, 1, 1], so the innermost dimension is even. However, the previous code determined it to be odd, which caused the weight reorder to be inserted.

Reply:
If it is the case I suggest you to have this adding pad only for allow_new_shape_infer is true, because 4D compressed weights are currently only observed on the dynamic models.

Reply:
Modified the code as discussed offline. thanks.